qadence_libs/qinfo_tools/qfi.py

```diff
@@ -0,0 +1,166 @@
+from __future__ import annotations
+
+from typing import Iterable
+
+import torch
+from qadence import Overlap
+from qadence.blocks import parameters, primitive_blocks
+from qadence.circuit import QuantumCircuit
+from qadence.types import BackendName, DiffMode, OverlapMethod
+from torch import Tensor
+
+from qadence_libs.qinfo_tools.spsa import spsa_2gradient_step
+from qadence_libs.qinfo_tools.utils import hessian
+
+
+def _symsqrt(A: Tensor) -> Tensor:
+    """Computes the square root of a Symmetric or Hermitian positive definite matrix.
+
+    or batch of matrices.
+
+    Code from https://github.com/pytorch/pytorch/issues/25481#issuecomment-1032789228
+    """
+    L, Q = torch.linalg.eigh(A)
+    zero = torch.zeros((), device=L.device, dtype=L.dtype)
+    threshold = L.max(-1).values * L.size(-1) * torch.finfo(L.dtype).eps
+    L = L.where(L > threshold.unsqueeze(-1), zero)  # zero out small components
+    return (Q * L.sqrt().unsqueeze(-2)) @ Q.mH
+
+
+def _set_circuit_vparams(circuit: QuantumCircuit, vparams_values: Iterable) -> None:
+    """Sets the variational parameter values of the circuit."""
+    if vparams_values is not None:
+        blocks = primitive_blocks(circuit.block)
+        iter_index = iter(range(len(blocks)))
+        for block in blocks:
+            params = parameters(block)
+            for p in params:
+                if p.trainable:
+                    p.value = float(vparams_values[next(iter_index)])  # type: ignore
+
+
+def _get_fm_dict(circuit: QuantumCircuit) -> dict:
+    """Returns a dictionary holding the FM parameters of the circuit."""
+    fm_dict = {}
+    blocks = primitive_blocks(circuit.block)
+    for block in blocks:
+        params = parameters(block)
+        for p in params:
+            if not p.trainable:
+                fm_dict[p.name] = torch.tensor([p.value])
+    return fm_dict
+
+
+def get_quantum_fisher(
+    circuit: QuantumCircuit,
+    vparams_values: Iterable[float | Tensor] | None = None,
+    fm_dict: dict[str, Tensor] = {},
+    backend: BackendName = BackendName.PYQTORCH,  # type: ignore
+    overlap_method: OverlapMethod = OverlapMethod.EXACT,
+    diff_mode: DiffMode = DiffMode.AD,  # type: ignore
+) -> Tensor:
+    """Returns the exact Quantum Fisher Information (QFI) matrix of the quantum circuit.
+
+    with given values for the variational parameters (vparams_values) and the
+    feature map (fm_dict).
+
+    Args:
+        circuit (QuantumCircuit): The Quantum circuit we want to compute the QFI matrix of.
+        vparams (tuple | list | Tensor | None):
+            Values of the variational parameters where we want to compute the QFI.
+        fm_dict (dict[str, Tensor]): Values of the feature map parameters.
+        overlap_method (OverlapMethod, optional): Defaults to OverlapMethod.EXACT.
+        diff_mode (DiffMode, optional): Defaults to DiffMode.ad.
+    """
+
+    # Get feature map dictionary (required to run Overlap().forward())
+    if fm_dict == {}:
```

Comment:
Are you checking for emptiness ? In case, I'll do:

```suggestion
    if not fm_dict:
```